lib/google/gax/configuration.rb

```diff
@@ -0,0 +1,593 @@
+    #   config.opt1 = "hi"      #=> "hi" (but prints a warning)
+    #   config.opt1 = nil       #=> nil (but prints a warning)
+    #
+    #   config.opt2             #=> :one
+    #   config.opt2 = :two      #=> :two
+    #   config.opt2             #=> :two
+    #   config.opt2 = :four     #=> :four (but prints a warning)
+    #
+    #   config.opt3             #=> "hi"
+    #   config.opt3 = "hiho"    #=> "hiho"
+    #   config.opt3             #=> "hiho"
+    #   config.opt3 = "HI"      #=> "HI" (but prints a warning)
+    #
+    #   config.opt4             #=> "yo"
+    #   config.opt4 = :yo       #=> :yo (Strings and Symbols allowed)
+    #   config.opt4             #=> :yo
+    #   config.opt4 = 3.14      #=> 3.14 (but prints a warning)
+    #   config.opt4 = nil       #=> nil (no warning: nil allowed)
+    #
+    #   config.sub              #=> <Google::Gax::Configuration>
+    #
+    #   config.sub.opt5         #=> false
+    #   config.sub.opt5 = true  #=> true  (true and false allowed)
+    #   config.sub.opt5         #=> true
+    #   config.sub.opt5 = nil   #=> nil (but prints a warning)
+    #
+    #   config.opt9 = "hi"      #=> "hi" (warning about unknown key)
+    #   config.opt9             #=> "hi" (no warning: key now known)
+    #   config.sub.opt9         #=> nil (warning about unknown key)
+    #
+    class Configuration < BasicObject
+      ##
+      # Constructs a Configuration object. If a block is given, yields `self` to the
+      # block, which makes it convenient to initialize the structure by making
+      # calls to `add_field!` and `add_config!`.
+      #
+      # @param [boolean] show_warnings Whether to print warnings when a
+      #     validation fails. Defaults to `true`.
+      # @return [Configuration] The constructed Configuration object.
+      #
+      def self.create show_warnings: true
+        config = new [], show_warnings: show_warnings
+        yield config if block_given?
+        config
+      end
+
+      ##
+      # Determines if the given object is a config. Useful because Configuration
+      # does not define the `is_a?` method.
+      #
+      # @return [boolean]
+      #
+      def self.config? obj
+        Configuration.send :===, obj
+      end
+
+      ##
+      # Internal constructor. Generally you should not call `new` directly,
+      # but instead use the `Configuration.create` method. The initializer is used
+      # directly by a few older clients that expect a legacy interface.
+      #
+      # @private
+      #
+      def initialize legacy_categories = {}, opts = {}
+        @show_warnings = opts.fetch :show_warnings, false
+        @values = {}
+        @defaults = {}
+        @validators = {}
+        add_options legacy_categories
+      end
+
+      ##
+      # Legacy method of adding subconfigs. This is used by older versions of
+      # the stackdriver client libraries but should not be used in new code.
+      #
+      # @deprecated
```

Comment:
Can we just remove these deprecated/legacy methods since we don't have existing users of this copy of the class?

(If we remove, also remove `:add_options` from `ILLEGAL_KEYS` below.)

Reply:
LOL, I just copied/pasted because that's who I am.

Reply:
"Good developers copy. Great developers paste." —Kelsey Hightower 🤣